common/logisticspipes/utils/GuideBookContents.kt

```diff
@@ -8,56 +9,84 @@ import net.minecraft.util.ResourceLocation
-      }
-    } catch (e: IOException) {
-      LogisticsPipes.log.error("Couldn't find page $index for language '$lang'!")
-    }
-    return null
-  }
-
-  companion object {
-
-    private val rm = Minecraft.getMinecraft().resourceManager
-
-    @JvmStatic
-    @JvmOverloads
-    fun load(lang: String = Minecraft.getMinecraft().languageManager.currentLanguage.languageCode): GuideBookContents? {
-      try {
-        val res = rm.getResource(ResourceLocation(LPConstants.LP_MOD_ID, "book/$lang/book.json"))
-        res.use {
-          val json = JsonParser().parse(InputStreamReader(res.inputStream)).asJsonObject
-          val title = json.get("title").asString
-          val pages = json.get("page_count").asInt
-          if (pages < 1) {
-            LogisticsPipes.log.error("Guide book for language '$lang' has $pages pages?!")
+class GuideBookContents private constructor(val lang: String, val divisions: List<Division> = mutableListOf(), val title: String) {
+
+    companion object {
+
+        private val rm = Minecraft.getMinecraft().resourceManager
+
+
+        @JvmStatic
+        @JvmOverloads
+        fun load(lang: String = Minecraft.getMinecraft().languageManager.currentLanguage.languageCode): GuideBookContents? {
+            try {
+                val res = rm.getResource(ResourceLocation(LPConstants.LP_MOD_ID, "book/$lang/book.json"))
+                res.use {
+                    val json = JsonParser().parse(InputStreamReader(res.inputStream)).asJsonObject
+                    val title = json.get("title").asString
+                    val divisions = json.get("divisions").asJsonArray.withIndex().map { (sIndex, divs) -> Division(lang, divs.asJsonObject, sIndex) }
+                    return GuideBookContents(lang, divisions, title)
+                }
+            } catch (e: IOException) {
+                if (lang != "en_us") {
+                    LogisticsPipes.log.error("Couldn't find book data for language '$lang'! Falling back to en_us…")
+                    return load("en_us")
+                } else {
+                    LogisticsPipes.log.error("Couldn't find book data for en_us! Something is horribly wrong…")
+                }
+            } catch (e: UnsupportedOperationException) {
+                LogisticsPipes.log.error("Invalid book.json for language '$lang'!", e)
+            }
             return null
-          }
-          return GuideBookContents(lang, pages, title)
         }
-      } catch (e: IOException) {
-        if (lang != "en_us") {
-          LogisticsPipes.log.error("Couldn't find book data for language '$lang'! Falling back to en_us…")
-          return load("en_us")
-        } else {
-          LogisticsPipes.log.error("Couldn't find book data for en_us! Something is horribly wrong…")
+
+
+    }
+
+    fun getDivision(index: Int): Division{
+        return divisions[index]
+    }
+
+    fun getChapter(dindex: Int, index: Int): Chapter?{
+        return getDivision(dindex).getChapter(index)
```

Comment:
Why are these needed? Just call getDivision().getChapter(), it's not really more code and is more readable

Reply:
If `getDivision().getChapter()` is more readable I guess they aren't needed, just thought having a function specifically for that purpose would be more readable, I'll remove then

Reply:
In this case you have 2 parameters, which you won't know what they do without looking at the parameter names or the function body, and when not using this function you'd have only 1 parameter but 2 calls, but since the function names are stuff like `getDivision` it's clear what the parameter is

Reply:
I changed this but I'll show you why I had that in the first place:
`currentPage = new SavedTab(gbc.getDivision(nbtTagCompound.getInteger("division")).getChapter(nbtTagCompound.getInteger("chapter")).getPage(nbtTagCompound.getInteger("page")), page, nbtTagCompound.getFloat("sliderProgress"));`

Reply:
Yeah, that's a very long line. In that case, I'd just insert some line breaks to make it look better.
```java
currentPage = new SavedTab(
  gbc.getDivision(nbtTagCompound.getInteger("division"))
    .getChapter(nbtTagCompound.getInteger("chapter"))
    .getPage(nbtTagCompound.getInteger("page")),
  page, nbtTagCompound.getFloat("sliderProgress")
);
```
Something like that.

Reply:
Yes I better do that :)